### PR TITLE
feat: new endpoint to validate ACL status

### PR DESCRIPTION
**Pull request by gabrielfs7**
Debugging ACL is a pain for developers, it is really tricky to collect all the information and we need several times to put logs manually in the server to understand the access issues. 

We have already several ACL tasks that are a nightmare to debug and took almost 1 week to be done. Example: https://oat-sa.atlassian.net/browse/ADF-1020

Motivated by this, I decided to introduce an endpoint that will list to us all the ACL status for a given Controller::Action, user and resource, all the the same time. Example: 

Parameters

- controller = controller FQN
- action = controller method name
- userId = the target user id
- requestParameters = The request parameters used in the target endpoint, so we can simulate taoDacSimple access validation

Calling URL providing an endpoint that contains taoDacSimple, funcAcl and ActionAccess restriction, we can get all the access restrictions and permissions listed.

```
https://test-tao-deploy.docker.localhost/funcAcl/AclStatus/check?controller=taoItems_actions_Items&action=editClassLabel&userId=https%3A%2F%2Ftest-tao-deploy.docker.localhost%2Fontologies%2Ftao.rdf%23superUser&requestParameters[id]=http%3A%2F%2Fwww.tao.lu%2FOntologies%2FTAOItem.rdf%23Item
```

Response: 

```
{
   "controller":"taoItems_actions_Items",
   "action":"editClassLabel",
   "requestParameters": {
        "id": "http://www.tao.lu/Ontologies/TAOItem.rdf#Item"
   },
   "user":{
      "id":"https:\/\/test-tao-deploy.docker.localhost\/ontologies\/tao.rdf#superUser",
      "class":"core_kernel_users_GenerisUser",
      "roles":[
         "http:\/\/www.tao.lu\/Ontologies\/TAO.rdf#GlobalManagerRole",
         "http:\/\/www.tao.lu\/Ontologies\/generis.rdf#taoPublishingManager",
         "http:\/\/www.tao.lu\/Ontologies\/TAO.rdf#BackOfficeRole",
         "http:\/\/www.tao.lu\/Ontologies\/TAO.rdf#BaseUserRole",
         "http:\/\/www.tao.lu\/Ontologies\/generis.rdf#AnonymousRole",
         "http:\/\/www.tao.lu\/Ontologies\/generis.rdf#GenerisRole",
         "http:\/\/www.tao.lu\/Ontologies\/generis.rdf#taoDeliverConnectManager",
         "http:\/\/www.tao.lu\/Ontologies\/generis.rdf#taoLtiConsumerManager",
         "http:\/\/www.tao.lu\/Ontologies\/TAOGroup.rdf#GroupsManagerRole",
         "http:\/\/www.tao.lu\/Ontologies\/TAOItem.rdf#ItemsManagerRole",
         "http:\/\/www.tao.lu\/Ontologies\/TAOItem.rdf#QTIManagerRole",
         "http:\/\/www.tao.lu\/Ontologies\/generis.rdf#parccTeiManager",
         "http:\/\/www.tao.lu\/Ontologies\/generis.rdf#qtiItemPciManager",
         "http:\/\/www.tao.lu\/Ontologies\/TAOLTI.rdf#LtiManagerRole",
         "http:\/\/www.tao.lu\/Ontologies\/TAOLTI.rdf#LtiDeliveryProviderManagerRole",
         "http:\/\/www.tao.lu\/Ontologies\/TAOMedia.rdf#MediaManagerRole",
         "http:\/\/www.tao.lu\/Ontologies\/TAOProctor.rdf#TestCenterManager",
         "http:\/\/www.tao.lu\/Ontologies\/TAOResultServer.rdf#ResultServerRole",
         "http:\/\/www.tao.lu\/Ontologies\/TAOResult.rdf#ResultsManagerRole",
         "http:\/\/www.tao.lu\/Ontologies\/TAOSubject.rdf#SubjectsManagerRole",
         "http:\/\/www.tao.lu\/Ontologies\/TAOTest.rdf#TaoQtiManagerRole",
         "http:\/\/www.tao.lu\/Ontologies\/TAOTest.rdf#TaoQtiTestPreviewerRole",
         "http:\/\/www.tao.lu\/Ontologies\/TAOTestCenter.rdf#TestCenterManager",
         "http:\/\/www.tao.lu\/Ontologies\/TAOTest.rdf#TestsManagerRole",
         "http:\/\/www.tao.lu\/Ontologies\/TAO.rdf#TaoDevToolsRole",
         "http:\/\/www.tao.lu\/Ontologies\/TAO.rdf#TaoManagerRole",
         "http:\/\/www.tao.lu\/Ontologies\/generis.rdf#ltiProctoringManager",
         "http:\/\/www.tao.lu\/Ontologies\/generis.rdf#pciSamplesManager",
         "http:\/\/www.tao.lu\/Ontologies\/generis.rdf#taoBackOfficeManager",
         "http:\/\/www.tao.lu\/Ontologies\/generis.rdf#taoCeManager",
         "http:\/\/www.tao.lu\/Ontologies\/generis.rdf#taoClientDiagnosticManager",
         "http:\/\/www.tao.lu\/Ontologies\/generis.rdf#taoColorContrastManager",
         "http:\/\/www.tao.lu\/Ontologies\/generis.rdf#taoDacSimpleManager",
         "http:\/\/www.tao.lu\/Ontologies\/generis.rdf#taoEventLogManager",
         "http:\/\/www.tao.lu\/Ontologies\/generis.rdf#taoDeliveryRdfManager",
         "http:\/\/www.tao.lu\/Ontologies\/generis.rdf#taoRevisionManager",
         "http:\/\/www.tao.lu\/Ontologies\/generis.rdf#taoStylesManager",
         "http:\/\/www.tao.lu\/Ontologies\/generis.rdf#taoTaskQueueManager",
         "http:\/\/www.tao.lu\/Ontologies\/generis.rdf#taoTestRunnerPluginsManager",
         "http:\/\/www.tao.lu\/Ontologies\/generis.rdf#taoTextHelpManager",
         "http:\/\/www.tao.lu\/Ontologies\/taoFuncACL.rdf#FuncAclManagerRole",
         "http:\/\/www.tao.lu\/Ontologies\/generis.rdf#xmlEditManager",
         "http:\/\/www.tao.lu\/Ontologies\/TAO.rdf#SysAdminRole",
         "http:\/\/www.tao.lu\/Ontologies\/TAO.rdf#DacAdministrator"
      ]
   },
   "funcAclPermissions":{
      "accessAllowed":true,
      "controllerAllowedRoles":[
         "http:\/\/www.tao.lu\/Ontologies\/TAOItem.rdf#ItemsManagerRole"
      ],
      "actionAllowedRoles":[
         "http:\/\/www.tao.lu\/Ontologies\/TAOItem.rdf#ItemClassNavigatorRole"
      ]
   },
   "taoActionAccessPermissions":{
      "accessAllowed":{
         "READ":true,
         "WRITE":true,
         "GRANT":true
      },
      "allowedPermissions":{
         "http:\/\/www.tao.lu\/Ontologies\/TAOItem.rdf#ItemClassNavigatorRole":"READ",
         "http:\/\/www.tao.lu\/Ontologies\/TAOItem.rdf#ItemClassEditorRole":"WRITE"
      }
   },
   "resourcePermissions":{
      "requiredRights":null,
      "allowedAccess":true
   }
}
```

### TODO 

- [ ] Allow endpoint access only to Admin user.
- [ ] Tests (if possible, since this feature aggregates value for developers debug and we have no budget to work too much on this).
- [ ] Use DI container instead of ServiceManager and ConfigurableService 
- [ ] Add README.md or api docs to document how to call this endpoint.
- [ ] Add other permissions source like TaoResourceWorkflow, etc

### CODE REVIEW

File: models/AclStatusCheckService.php [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+        if (!isset($filter['userId'], $filter['controller'], $filter['action'])) {
+            throw new InvalidArgumentException('Missing required parameters');
+        }
+
+        $user = $this->userService->getUser($filter['userId']);
+        $controller = $filter['controller'];
+        $action = $filter['action'];
+        $requestParameters = $filter['requestParameters'] ?? [];
+        $userRoles = $this->getUserRoles($user);
+
```
Comment (hectoras) [inline models/AclStatusCheckService.php:105]: suggestion: If all these array indexes are mandatory, maybe its values can be passed as standard parameters instead.

```suggestion
    public function check(string $userId, string $controller, string $action, array $requestParameters = []): array
    {
        if (!isset($userId, $controller, $action)) {
            throw new InvalidArgumentException('Missing required parameters');
        }

        $user = $this->userService->getUser($userId);
        $userRoles = $this->getUserRoles($user);

```

**Review by hectoras — APPROVED** [review]

File: models/AclStatusCheckService.php (hunk 1 above)
Comment (rgomeztao) [reply models/AclStatusCheckService.php:105]: :+1:

File: models/AclStatusCheckService.php [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+         */
+        $this->userService = ServiceManager::getServiceManager()->get(UserService::SERVICE_ID);
+        $this->usersService = core_kernel_users_Service::singleton();
+        $this->actionAccessControl = ServiceManager::getServiceManager()->get(ActionAccessControl::SERVICE_ID);
+        $this->parameterFilter = new ParameterFilterProxy();
+        $this->dataAccessControl = new DataAccessControl();
+    }
+
+    public function overloadControllerAccess(array $controllerAccess): self
+    {
```
Comment (rgomeztao) [inline models/AclStatusCheckService.php:81]: do we expect public functions in a controller? :thinking:

**Review by rgomeztao — APPROVED** [review]
Approved, left some comments :)

File: models/AclStatusCheckService.php [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2022 (original work) Open Assessment Technologies SA;
```
[the remark this answers is not in the document]
Comment (gabrielfs7) [reply models/AclStatusCheckService.php:18]: Look when I opened this PR ⏳ :)

**Pull request closed without merging**